serviceregistration/kubernetes/retry_handler.go

```diff
@@ -0,0 +1,112 @@
+package kubernetes
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/serviceregistration/kubernetes/client"
+)
+
+// How often to retry sending a state update if it fails.
+var retryFreq = 5 * time.Second
+
+// retryHandler executes retries.
+// It is thread-safe.
+type retryHandler struct {
+	// These don't need a mutex because they're never mutated.
+	logger             hclog.Logger
+	namespace, podName string
+	client             *client.Client
+
+	// This gets mutated on multiple threads.
+	patchesToRetry     []*client.Patch
+	patchesToRetryLock sync.Mutex
+}
+
+// Run runs at an interval, checking if anything has failed and if so,
+// attempting to send them again.
+func (r *retryHandler) Run(shutdownCh <-chan struct{}, wait *sync.WaitGroup) {
+	// Make sure Vault will give us time to finish up here.
+	wait.Add(1)
+	defer wait.Done()
+
+	retry := time.NewTicker(retryFreq)
```

Comment:
Where is this ticker being shut down?

Reply:
Nice catch, will add.

Reply:
`time.Sleep` can avoid the additional code to manage the ticker.

Reply:
Added a commit stopping the ticker: https://github.com/hashicorp/vault/pull/8249/commits/03573a8cd1579050a9f9e9ec39e1ebbd2723eb78

Regarding using `time.Sleep`, the reason I didn't use it here is because it's not interruptible should the `shutdownCh` be closed. It just hangs. If I used it and the application were shutting down, it would hang until the sleep finished. I also considered replacing the `<-retry.C` case with `<-time.NewTimer(retryFreq).C`, but decided against it because I don't like the idea of instantiating something on each iteration of an infinite loop. Instantiating the ticker outside the loop and then continuing to use it seems safer to me.